web/src/main/java/com/google/accompanist/web/WebView.kt

```diff
@@ -0,0 +1,164 @@
+                        // It's important we still support a link to about:blank though.
+                        if (webViewLoaded || url != "about:blank") {
+                            onContentChanged(WebContent.Url(Uri.parse(url)))
+                            webViewLoaded = true
+                        }
+                    }
+
+                    override fun onPageFinished(view: WebView?, url: String?) {
+                        super.onPageFinished(view, url)
+                        state.isLoading = false
+                    }
+
+                    override fun shouldOverrideUrlLoading(
+                        view: WebView?,
+                        request: WebResourceRequest?
+                    ): Boolean {
+                        // Override all url loads to make the single source of truth
+                        // of the URL the state holder Url
+                        request?.let {
+                            val content = WebContent.Url(it.url)
+                            onContentChanged(content)
+                        }
+                        return true
+                    }
+                }
+            }.also { view = it }
+        },
+        modifier = modifier
+    ) { webView ->
+        when (val l = state.content) {
+            is WebContent.Url -> {
+                val url = l.uri.toString()
+
+                if (url.isNotEmpty() && url != webView.url) {
+                    webView.loadUrl(url)
+                }
+            }
+            is WebContent.Data -> {
+                webView.loadDataWithBaseURL(l.baseUrl, l.data, null, null, null)
+            }
+        }
+
+        state.canGoBack = webView.canGoBack()
+    }
+}
+
+sealed class WebContent {
+    data class Url(val uri: Uri) : WebContent()
+    data class Data(val data: String, val baseUrl: String? = null) : WebContent()
+}
+
+/**
+ * A state holder to hold the state for the WebView. In most cases this will be remembered
+ * using the rememberWebViewState(uri) function.
+ */
+class WebViewState(webContent: WebContent) {
+    /**
+     *  The content being loaded by the WebView
+     */
+    var content by mutableStateOf<WebContent>(webContent)
+
+    /**
+     * Whether the WebView is currently loading data in it's main frame
+     */
+    var isLoading: Boolean by mutableStateOf(false)
+        internal set
+
+    /**
+     * Whether the WebView can navigate back
+     */
+    var canGoBack: Boolean by mutableStateOf(false)
+        internal set
+}
+
+/**
+ * Creates a WebView state that is remembered across Compositions.
+ *
+ * @param uri The uri to load in the WebView
+ */
+@Composable
+fun rememberWebViewState(uri: Uri) =
+    remember { WebViewState(WebContent.Url(uri)) }
+
+/**
+ * Creates a WebView state that is remembered across Compositions.
+ *
+ * @param data The uri to load in the WebView
+ */
+@Composable
+fun rememberWebViewState(data: String) =
```

Comment:
What about the optional `baseUrl`? Should that be passed in here as well?

Reply:
Good spotting, yes. Will fix